pieces/shared/menu-button/container.tsx

```diff
@@ -0,0 +1,170 @@
+};
+
+export const MenuButton = ({ label, children }: { label: string; children: ReactNode }) => {
+  const [isOpen, toggleIsOpen] = useReducer((isOpen) => !isOpen, false);
+
+  const [initialFocus, setInitialFocus] = useState<'first' | 'last'>('first');
+
+  const menuRef = useRef<MenuHandle>(null);
+
+  const triggerRef = useRef<HTMLButtonElement>(null);
+
+  const shouldResetFocusRef = useRef(false);
+
+  useEffect(() => {
+    if (!isOpen && shouldResetFocusRef.current) {
+      shouldResetFocusRef.current = false;
+      window.setTimeout(() => {
+        triggerRef.current?.focus();
+      }, 0);
+    }
+  }, [isOpen]);
+
+  const onKeyDown: KeyboardEventHandler<HTMLButtonElement> = useCallback((event) => {
+    switch (event.key) {
+      case 'ArrowDown':
+        event.preventDefault();
+        toggleIsOpen();
+        setInitialFocus('first');
+        break;
+
+      case 'ArrowUp':
+        event.preventDefault();
+        toggleIsOpen();
+        setInitialFocus('last');
+        break;
+    }
+  }, []);
+
+  useEffect(() => {
+    if (isOpen === false) {
+      setInitialFocus('first');
+      return;
+    }
+
+    if (initialFocus === 'first') {
+      menuRef.current?.focusFirst();
+    }
+
+    if (initialFocus === 'last') {
+      menuRef.current?.focusLast();
+    }
+  }, [initialFocus, isOpen]);
+
+  const containerRef = useRef<HTMLSpanElement>(null);
+
+  const onMenuKeyDown: KeyboardEventHandler<HTMLButtonElement> = useCallback((event) => {
+    switch (event.key) {
+      case 'Escape':
+        shouldResetFocusRef.current = true;
+        toggleIsOpen();
+        return;
+
+      case 'ArrowDown':
+        menuRef.current?.focusNext();
+        event.preventDefault();
+        return;
+
+      case 'ArrowUp':
+        menuRef.current?.focusPrev();
+        event.preventDefault();
+        break;
+
+      case 'Home':
+        menuRef.current?.focusFirst();
+        break;
+
+      case 'End':
+        menuRef.current?.focusLast();
+        break;
+    }
+
+    if (containerRef.current && /[a-z]/i.test(event.key)) {
+      const letter = event.key.toLowerCase();
+
+      const items = Array.from(containerRef.current.querySelectorAll('li'));
+
+      const activeIndex = items.indexOf(document.activeElement as HTMLLIElement);
+
+      let match = items
+        .slice(activeIndex + 1)
+        .find((item) => item.textContent?.toLowerCase().startsWith(letter));
+
+      if (match === undefined) {
+        match = items.find((item) => item.textContent?.toLowerCase().startsWith(letter));
+      }
+
+      if (match !== undefined) {
+        match.focus();
+      }
+    }
+  }, []);
+
+  useEffect(() => {
+    if (!isOpen) {
+      return;
+    }
+
+    window.addEventListener('click', toggleIsOpen);
+
+    return () => {
+      window.removeEventListener('click', toggleIsOpen);
+    };
+  }, [isOpen]);
+
+  const onClick: MouseEventHandler = useCallback((event) => {
+    toggleIsOpen();
+    event.stopPropagation();
```

Comment:
why is the event stopped rather than prevented?

Reply:
I don't remember exactly why I was doing this, but have removed it now